grpclb/src/main/java/io/grpc/grpclb/GrpclbLoadBalancer.java

```diff
@@ -0,0 +1,323 @@
+    synchronized (lock) {
+      lastError = statusException;
+      pendingPicksFulfillmentBatch = pendingPicks.createFulfillmentBatch();
+    }
+    pendingPicksFulfillmentBatch.fail(statusException);
+  }
+
+  private class LbResponseObserver implements StreamObserver<LoadBalanceResponse> {
+    @Override public void onNext(LoadBalanceResponse response) {
+      logger.info("Got a LB response: " + response);
+      FulfillmentBatch<ClientTransport> pendingPicksFulfillmentBatch;
+      final RoundRobinServerList newRoundRobinServerList;
+      HashMap<SocketAddress, ResolvedServerInfo> newServerMap =
+          new HashMap<SocketAddress, ResolvedServerInfo>();
+      synchronized (lock) {
+        InitialLoadBalanceResponse initialResponse = response.getInitialResponse();
+        // TODO(zhangkun83): make use of initialResponse
+        RoundRobinServerList.Builder listBuilder = new RoundRobinServerList.Builder(tm);
+        ServerList serverList = response.getServerList();
+        // TODO(zhangkun83): honor expiration_interval
+        for (Server server : serverList.getServersList()) {
+          if (server.getDropRequest()) {
+            listBuilder.add(null);
+          } else {
+            InetSocketAddress address = new InetSocketAddress(
+                server.getIpAddress(), server.getPort());
+            listBuilder.add(address);
+            // TODO(zhangkun83): fill the LB token to the attributes, and insert it to the
+            // application RPCs.
+            if (!newServerMap.containsKey(address)) {
+              newServerMap.put(address, new ResolvedServerInfo(address, Attributes.EMPTY));
+            }
+          }
+        }
+        newRoundRobinServerList = listBuilder.build();
+        if (newRoundRobinServerList.size() == 0) {
+          return;
+        }
+        roundRobinServerList = newRoundRobinServerList;
+        servers = newServerMap;
+        pendingPicksFulfillmentBatch = pendingPicks.createFulfillmentBatch();
+      }
+      tm.updateRetainedTransports(newServerMap.keySet());
+      pendingPicksFulfillmentBatch.link(
+          new Supplier<ListenableFuture<ClientTransport>>() {
+            @Override
+            public ListenableFuture<ClientTransport> get() {
+              return newRoundRobinServerList.getTransportForNextServer();
+            }
+          });
+    }
+
+    @Override public void onError(Throwable error) {
+      onStreamClosed(Status.fromThrowable(error)
+          .augmentDescription("Stream to GRPCLB LoadBalancer had an error"));
+    }
+
+    @Override public void onCompleted() {
+      onStreamClosed(Status.UNAVAILABLE.augmentDescription(
+          "Stream to GRPCLB LoadBalancer was closed"));
+    }
+
+    private void onStreamClosed(Status status) {
+      handleError(status);
+      synchronized (this) {
+        if (lbResponseObserver == this) {
+          // I am still the active LB stream. Reopen the stream.
+          startNegotiation();
```

Comment:
If the server is available but not a load balancer, it seems this will spam the server continually with requests.